packages/button/src/button.tsx

```diff
@@ -1,35 +1,92 @@
 import React from 'react';
 import { yoru } from '@yoru-ui/core';
 import { useResolvedThemes } from '@yoru-ui/themes';
+import LoadingIcon from './LoadingIcon';
+import { AnchorButtonProps, ButtonProps, NativeButtonProps } from './button.interface';
 
-type ButtonThemeProps = {
-  variants?: 'solid' | 'outline' | 'ghost' | 'link';
-  colorScheme?: 'default' | 'primary' | 'secondary' | 'danger';
-  sizes?: 'sm' | 'md' | 'lg';
-};
-export interface ButtonDefaultProps extends ButtonThemeProps {
-  block?: boolean;
-  className?: string;
-  children?: React.ReactNode;
-}
-
-export const Button: React.FC<ButtonDefaultProps> = props => {
-  const { colorScheme = 'default', sizes = 'md', block = false, className, children } = props;
-
-  const buttonStyled = useResolvedThemes('Button', { colorScheme, sizes });
-
-  const buttonStyles = () => {
-    return {
-      ...buttonStyled,
-      width: block ? '100%' : 'auto',
-    };
-  };
+export const Button: React.FC<ButtonProps> = props => {
+  const {
+    variants = 'solid',
+    colorScheme = 'gray',
+    sizes = 'md',
+    block = false,
+    className,
+    children,
+    icon,
+    iconPosition = 'left',
+    loading = false,
+    disabled,
+    ...rest
+  } = props;
+
+  const buttonStyled = useResolvedThemes('Button', {
+    variants,
+    sizes,
+    colorScheme,
+  });
 
   const classes = (...classNames: any[]) => classNames.filter(Boolean).join(' ');
 
-  return (
-    <yoru.button type="button" __style={buttonStyles} className={classes('yoru-button', className)}>
+  const styles = {
+    ...buttonStyled,
+    width: block ? '100%' : 'auto',
+  };
+
+  const iconType = loading ? <LoadingIcon /> : <yoru.div>{icon}</yoru.div>;
+
+  const handleClick = (e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement, MouseEvent>) => {
+    const { onClick } = props;
+
+    if (loading || disabled) {
+      e.preventDefault();
+      return;
+    }
+    (onClick as React.MouseEventHandler<HTMLButtonElement | HTMLAnchorElement>)?.(e);
+  };
+
+  // set icon position always left when loading
+  const setIconPosition = iconPosition !== 'left' && !loading ? 'row-reverse' : 'initial';
+
+  const ButtonContent = () => (
+    <yoru.div
+      __style={{
+        display: 'flex',
+        alignItems: 'center',
+        columnGap: !icon && !loading ? '0' : '0.5rem',
+        flexDirection: setIconPosition,
+      }}
+    >
+      {iconType}
       {children}
-    </yoru.button>
+    </yoru.div>
+  );
+
+  if (variants === 'link') {
```

Comment:
karna props nya ini sama semua better dibuat gini sih
`const ButtonComponent = variants === 'link' ? yoru.a : yoru.button`

nanti dibawah tinggal di return sekali aja
` <ButtonComponent
        __style={styles}
        className={classes('yoru-button', className)}
        onClick={handleClick}
        disabled={disabled || loading}
        {...(rest as AnchorButtonProps)}
      >
        <ButtonContent />
      </ButtonComponent>`

Reply:
noted, that nice suggestion. I'll enhance it later.